Create institution functionality

IvanPPetrov02 opened the pull request:
done the controller presenter and usecase for the create institution functionality

IvanPPetrov02 committed:
Create institution functionality

done the controller presenter and usecase for the create institution functionality

IvanPPetrov02 committed:
update on the controller

updated the controller to check for every property as it was missed

FelipeCarillo commented on src/modules/create_institution/app/create_institution_controller.ts:
```diff
@@ -0,0 +1,54 @@
+                social_medias: institutionData.social_medias
+            });
+
+            await this.usecase.execute(institution);
```
U need to pass to usecase, the request.headers and request.body.body

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,38 @@
+        this.database_repo = database_repo;
+    }
+
+    public async execute(institution: Institution){
```
Here u check if the headers has the authorization, check if the user is authenticated, and them u build the Institution class and create!

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,38 @@
+import { IInstitutionRepo } from '../../../core/repositories/interfaces/IInstitutionRepo';
+
+export class CreateInstitutionUsecase {
+    public database_repo: IInstitutionRepo;
```
Here u need to import the IUserRepo too, because u need to check if the user is in our db.

IvanPPetrov02 committed:
Updated the create institution

FelipeCarillo commented on src/modules/create_institution/app/create_institution_controller.ts:
```diff
@@ -0,0 +1,58 @@
+                throw new MissingParameter("Headers");
+            }
+
+            const institutionData = request.body.body;
+
+            if (!institutionData.id || !institutionData.name || !institutionData.description || !institutionData.email || !institutionData.country || !institutionData.images || !institutionData.social_medias) {
+                throw new MissingParameter("InstitutionData");
+            }
+
+            const institution = new Institution({
+                id: institutionData.id,
+                name: institutionData.name,
+                description: institutionData.description,
+                email: institutionData.email,
+                country: institutionData.country,
+                images: institutionData.images,
+                social_medias: institutionData.social_medias
+            });
```
This logic of creating a entity Institution, need to be implemented in the usecase, the controller, u only check if the request.body is not null, and them pass to usecase

IvanPPetrov02 replied:
so the usecase should be responsible for creating the institution?

FelipeCarillo replied:
Yes, the use case is responsible for all the logic and rules! The controller is used only to verify if the needed request is valid and them, he passes the body and headers, for exemple, to the use case!

FelipeCarillo commented on src/modules/create_institution/app/create_institution_controller.ts:
```diff
@@ -0,0 +1,58 @@
+                social_medias: institutionData.social_medias
+            });
+
+            await this.usecase.execute(institution, request.headers);
```
Instead of passing institution, u pass the request.body

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,58 @@
+        if (!headers.authorization) {
+            throw new MissingParameter("Authorization");
+        }
+        
```
U need to decode the token of the headers.authorization and them take the user_id and make a get_user with it, check in other routes!

IvanPPetrov02 replied:
really not sure about this one as well, so could you please explain again with more details?

FelipeCarillo replied:
Yes, the Authorization is the encrypted user_id encrypted with jwt, so to get the user_id for verifying if the user is in our database, you need first to decode the encrypeted user_id and after pass the response (user_id decrypted) to get_user function!

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,58 @@
+        
+        const userExists = await this.user_repo.get_user(headers.authorization);
+        if (!userExists) {
+            throw new Error("User does not exist in our database");
```
The error is UserNotAuthentificated

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,58 @@
+            throw new Error("User does not exist in our database");
+        }
+
+        const user = await this.user_repo.get_user(headers.authorization);
+        if (!user) {
+            throw new Error("User is not authenticated");
+        }
```
duplicated

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,58 @@
+            throw new Error("User is not authenticated");
+        }
+
+        if (!institution.name) {
```
duplicated

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,58 @@
+        if (!userExists) {
+            throw new Error("User does not exist in our database");
+        }
+
```
After getting the user, check if his type is ADMIN or MODERATOR, else throw UserNotAllowed

IvanPPetrov02 committed:
some of the changes complete

I have completed some of the changes needed but need more info for the other things

IvanPPetrov02 committed:
user is validated and entity is created in usecase

FelipeCarillo commented on src/modules/create_institution/app/create_institution_presenter.ts:
```diff
@@ -0,0 +1,18 @@
+const institutionRepo = new InstitutionRepo();
+const userRepo = new UserRepo();
+
+const usecase = new CreateInstitutionUsecase(institutionRepo, userRepo);
```
Here u don't need to instance these repos, you need to instance the Repository class, like this:
`const repository = new Repository({ user_repo: true, activity_repo: true });`
And after that, u pass the **repository.UserRepo** and the **repository.InstitutionRepo** to the Usecase.

FelipeCarillo commented on src/modules/create_institution/app/create_institution_controller.ts:
```diff
@@ -0,0 +1,42 @@
+                throw new MissingParameter("Headers");
+            }
+
+            await this.usecase.execute(request.body, request.headers);
```
pass here the request.body.body!

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public async execute(institutionData: any, headers: any){
+        if (!headers.authorization) {
```
is Authorization, not authorization

FelipeCarillo commented on src/modules/create_institution/app/create_institution_usecase.ts:
```diff
@@ -0,0 +1,64 @@
+            throw new MissingParameter("Authorization");
+        }
+        
+        const decodedToken = jwt.verify(headers.authorization, process.env.JWT_SECRET as string);
+        const userId = decodedToken;
+        
+        
+        const userExists = await this.user_repo.get_user(headers.authorization);
+        if (!userExists || userExists.user_type !== UserTypeEnum.ADMIN && userExists.user_type !== UserTypeEnum.MODERATOR) {
+            throw new UserNotAuthenticated("User is not authenticated");
+        }
```
Check how u do the logic in this usecase: [repo](https://github.com/Instituto-Maua-de-Tecnologia/coil_mss/blob/creating-get-activity/src/modules/get_activity/app/get_activity_usecase.ts)

Lines 40 to 51

FelipeCarillo commented on src/modules/create_institution/app/create_institution_controller.ts:
```diff
@@ -0,0 +1,42 @@
+            if (!request) {
+                throw new MissingParameter("Request");
+            }
+            if (!request.body || !request.body.body) { 
```
Only verify the body, the body.body is in the usecase the verification!

IvanPPetrov02 committed:
fixed the validation, and presenter sends correct things now
